cpp/src/arrow/util/list_util.cc

```diff
@@ -0,0 +1,353 @@
+#include "arrow/array/data.h"
+#include "arrow/type.h"
+#include "arrow/type_traits.h"
+#include "arrow/util/bit_run_reader.h"
+#include "arrow/util/checked_cast.h"
+#include "arrow/util/list_util.h"
+#include "arrow/util/logging.h"
+#include "arrow/util/string.h"
+
+namespace arrow::list_util {
+
+namespace internal {
+
+namespace {
+
+using arrow::internal::checked_cast;
+using arrow::internal::ReverseSetBitRunReader;
+using arrow::internal::SetBitRunReader;
+
+/// \pre input.length() > 0 && input.null_count() != input.length()
+/// \param input A LIST_VIEW or LARGE_LIST_VIEW array
+template <typename offset_type>
+std::optional<int64_t> MinViewOffset(const ArraySpan& input) {
+  const uint8_t* validity = input.buffers[0].data;
+  const auto* offsets = input.GetValues<offset_type>(1);
+  const auto* sizes = input.GetValues<offset_type>(2);
+
+  // Make an access to the sizes buffer only when strictly necessary.
+#define MINIMIZE_MIN_VIEW_OFFSET(i)             \
+  auto offset = offsets[i];                     \
+  if (min_offset.has_value()) {                 \
+    if (offset < *min_offset && sizes[i] > 0) { \
+      if (offset == 0) {                        \
+        return 0;                               \
+      }                                         \
+      min_offset = offset;                      \
+    }                                           \
+  } else {                                      \
+    if (sizes[i] > 0) {                         \
+      if (offset == 0) {                        \
+        return 0;                               \
+      }                                         \
+      min_offset = offset;                      \
+    }                                           \
+  }
+
+  std::optional<offset_type> min_offset;
+  if (validity == nullptr) {
+    for (int64_t i = 0; i < input.length; i++) {
+      MINIMIZE_MIN_VIEW_OFFSET(i);
+    }
+  } else {
+    SetBitRunReader reader(validity, input.offset, input.length);
+    while (true) {
+      const auto run = reader.NextRun();
+      if (run.length == 0) {
+        break;
+      }
+      for (int64_t i = run.position; i < run.position + run.length; ++i) {
+        MINIMIZE_MIN_VIEW_OFFSET(i);
+      }
+    }
+  }
+  return min_offset;
+
+#undef MINIMIZE_MIN_VIEW_OFFSET
+}
+
+/// \pre input.length() > 0 && input.null_count() != input.length()
+/// \param input A LIST_VIEW or LARGE_LIST_VIEW array
+template <typename offset_type>
+int64_t MaxViewEnd(const ArraySpan& input) {
+  constexpr auto kInt64Max = std::numeric_limits<int64_t>::max();
+  const auto values_length = input.child_data[0].length;
+
+  const uint8_t* validity = input.buffers[0].data;
+  const auto* offsets = input.GetValues<offset_type>(1);
+  const auto* sizes = input.GetValues<offset_type>(2);
+
+  // Early-exit: 64-bit overflow detected. This is not possible on a valid list-view,
+  // but we return the maximum possible value to avoid undefined behavior.
```

Comment:
Unless there's a specific condition where this can be called on invalid data (for example when reading IPC data), we don't care that UB would happen if the list-view is invalid.

Reply:
This is a public API. I can't guarantee on what kind of data it will be called.

Reply:
Pushing a commit that removes this check. I think I'm finally understanding where we are strict about validations and where we can assume valid as pre-conditions.